cpp/src/arrow/compute/kernels/isin.cc

```diff
@@ -0,0 +1,244 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/kernels/isin.h"
+
+#include <algorithm>
+#include <cstdint>
+#include <cstring>
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "arrow/array.h"
+#include "arrow/buffer.h"
+#include "arrow/builder.h"
+#include "arrow/compute/context.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/util-internal.h"
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/hashing.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/macros.h"
+#include "arrow/util/string_view.h"
+#include "arrow/visitor_inline.h"
+
+namespace arrow {
+
+class MemoryPool;
+
+using internal::checked_cast;
+using internal::DictionaryTraits;
+using internal::HashTraits;
+
+namespace compute {
+
+#define CHECK_IMPLEMENTED(KERNEL, FUNCNAME, TYPE)                                       \
+  if (!KERNEL) {                                                                        \
+    return Status::NotImplemented(FUNCNAME, " not implemented for ", type->ToString()); \
+  }
+
+class HashBinaryKernelImpl: public BinaryKernel {
+
+    virtual Status Compute(FunctionContext* ctx, const ArrayData& left,
+                         const ArrayData& right, std::shared_ptr<ArrayData>& out) = 0;
+
+    public:
+    Status Call(FunctionContext* ctx, const Datum& left, const Datum& right,
+                                                    Datum* out) override {
+
+        DCHECK_EQ(Datum::ARRAY, right.kind());
+        DCHECK_EQ(Datum::ARRAY, left.kind());
+
+        const ArrayData& left_data = *left.array();
+        const ArrayData& right_data = *right.array();
+
+        std::shared_ptr<ArrayData> result;
+        RETURN_NOT_OK(Compute(ctx, left_data , right_data, result));
+
+        out->value = std::move(result);
+        return Status::OK();
+    }
+    std::shared_ptr<DataType> out_type() const override { return boolean(); }
+};
+
+template <typename Type, typename Scalar>
+class IsInKernel : public HashBinaryKernelImpl{
+public:
+    IsInKernel(const std::shared_ptr<DataType>& type, MemoryPool* pool){}
+
+    Status VisitNull(){
+        values.emplace_back ();
+        return Status::OK();
+    }
+
+    Status VisitValue(const Scalar& value){
+        values.emplace_back(value);
+        return Status::OK();
+    }
+
+    Status Compute(FunctionContext* ctx, const ArrayData& left, const ArrayData& right,
+                 std::shared_ptr<ArrayData>& out) override {
+
+        values.reserve(left.length);
+        memo_table_.reset(new MemoTable(0));
+        bool_builder_.Reset();
```

Comment:
Can you use: PrimitiveAllocatingUnaryKernel or PrimitiveAllocatingBinaryKernel (https://github.com/apache/arrow/blob/c822e466b6e88f67059fea2da4a9fb5e5bfa201a/cpp/src/arrow/compute/kernels/util-internal.h) instead of allocated here.